src/Engine/ProtoCore/DSASM/Heap.cs

```diff
@@ -416,7 +424,28 @@ public StackValue AllocatePointer(int size, MetaData metadata)
         private StackValue AllocateStringInternal(string str, bool isConstant)
         {
             int index;
-            if (!stringTable.TryGetPointer(str, out index))
+            if (stringTable.TryGetPointer(str, out index))
+            {
+                // Any existing heap elements, marked as white, that are in the sweepSet and that are referenced during the sweep cycle will be marked as Black.
+                // This will ensure that no reachable data is mistakenly cleaned up.
+                bool isDuringGCCriticalAsyncCycle = gcState != GCState.Pause;// Between the time the GC takes a snapshot of the stack and heap untill GC cycle is over.
+                bool isValidHeapIndex = index >= 0 && index < heapElements.Count;
+                if (isDuringGCCriticalAsyncCycle && isValidHeapIndex)
+                {
+                    var he = heapElements[index];
+                    Validity.Assert(he != null, $"Null heap element found at index {index} during AllocateStringInternal");
```

Comment:
Sorry for nitpicking: here we are asserting that something cannot be null so I think it will be appropriate to negate the above message and make it sound something like: `Heap element found at index {index} during Allocate cannot be null`